tests_e2e/tests/publish_hostname/publish_hostname.py

```diff
@@ -0,0 +1,200 @@
+                                                                              self._context.vm,
+                                                                              self._context.vm.resource_group,
+                                                                              self._context.vm.subscription,
+                                                                              self._context.username,
+                                                                              self._vm_password))
+
+    def retry_ssh_if_connection_reset(self, command: str, use_sudo=False):
+        # The agent may bring the network down and back up to publish the hostname, which can reset the ssh connection.
+        # Adding retry here for connection reset.
+        retries = 3
+        while retries > 0:
+            try:
+                return self._ssh_client.run_command(command, use_sudo=use_sudo)
+            except CommandError as e:
+                retries -= 1
+                retryable = e.exit_code == 255 and "Connection reset by peer" in e.stderr
+                if not retryable or retries == 0:
+                    raise
+                log.warning("The SSH operation failed, retrying in 30 secs")
+                sleep(30)
+
+    def run(self):
+        # Add password to VM and log. This allows us to debug with serial console if necessary
+        self.add_vm_password()
+
+        # This test looks up what hostname is published to dns. Check that the tools necessary to get hostname are
+        # installed, and if not install them.
+        lookup_cmd, dns_regex = self.check_and_install_dns_tools()
+
+        # Enable agent hostname monitoring
+        log.info("Executing script update-waagent-conf to enable agent hostname monitoring")
+        result = self._ssh_client.run_command("update-waagent-conf Provisioning.MonitorHostName=y "
+                                              "Provisioning.MonitorHostNamePeriod=30", use_sudo=True)
+        log.info("Successfully enabled agent hostname monitoring config flag: {0}".format(result))
+
+        hostname_change_ctr = 0
+        # Update the hostname 3 times
+        while hostname_change_ctr < 3:
+            try:
+                hostname = "lisa-hostname-monitor-{0}".format(hostname_change_ctr)
+                log.info("Update hostname to {0}".format(hostname))
+                self.retry_ssh_if_connection_reset("hostnamectl set-hostname {0}".format(hostname), use_sudo=True)
+
+                # Wait for the agent to detect the hostname change for up to 2 minutes
+                timeout = datetime.datetime.now() + datetime.timedelta(minutes=2)
+                hostname_detected = ""
+                while datetime.datetime.now() <= timeout:
+                    try:
+                        hostname_detected = self.retry_ssh_if_connection_reset("grep -n {0} /var/log/waagent.log".format(hostname), use_sudo=True)
```

Comment:
can we add a little more context to grep's regex (on top of the hostname). The message is something like "Detected hostname change, etc"

Reply:
updated